### PR TITLE
Fleet RBAC - InheritedFleetWorkspacePermissions validation

### PR DESCRIPTION
**Pull request by raulcabello**
## Problem
Need to add validation for the new fields in `GlobalRole` `InheritedFleetWorkspacePermissions`

## Solution
- Validate the user have enough permission to create/update the rules defined in `InheritedFleetWorkspacePermissions.ResourceRules`
- Validate the user have enough permission to create/update the rules that are generated based on the `InheritedFleetWorkspacePermissions.WorkspaceVerbs`
- Don't allow to modify the labels `authz.management.cattle.io/gr-owner` and `authz.management.cattle.io/grb-owner` in `ClusterRole` and `ClusterRoleBinding`

## CheckList
  
- [x] Test
  
- [x] Docs

### CODE REVIEW

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (MbolotSuse) [inline pkg/resources/management.cattle.io/v3/globalrole/validator.go:1]: Please update the docs for this check as well.

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator_test.go
Comment (MbolotSuse) [inline pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator_test.go:61]: Would be good to add a test for the error case (where it can't pull the old/new rule from the request). Example [here](https://github.com/rancher/webhook/blob/8e3d96998fe1d95e71196c06d8c6dca26e284282/pkg/resources/core/v1/secret/validator_test.go#L212)

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go
Comment (MbolotSuse) [inline pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go:62]: Similar comment about handling the error case.

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (MbolotSuse) [inline pkg/resources/management.cattle.io/v3/globalrole/validator.go:175]: Right now, this seems to be checking if the user has the requested permission in each of the workspaces as a means of escalation checks. 

So for example, let's assume there are 2 fleetworkspaces: `fleet-default`, and `fleet-local`. If I'm trying to edit a global role that has the following permissions:
```yaml
apiVersion: management.cattle.io/v3
kind: GlobalRole
metadata:
  name: example
inheritedFleetWorkspacePermissions:
  resourceRules:
  - apiGroups:
    - ""
    resources: 
    - "pods"
    verbs: 
    -"get"
```
The webhook will check if I have `get` on `pods` in `fleet-default`, which I could already have by virtue of a kubernetes role/binding. The issue is, if someone makes a new workspace `fleet-new` I also gain permissions in that workspace. But the original role/binding didn't give me that new permission - so I've technically escalated my privileges.

In short, currently having permissions in all non-local namespaces is not a gaurantee that you will have them in all future workspaces as well. This issue also applies for the verb check below. 

I would recommend looking at how we do the escalation check for `inheritedClusterRoles` [here](https://github.com/rancher/webhook/blob/109fe534b73bc47374dfd5f8a962d3f3d6545eb6/pkg/resources/management.cattle.io/v3/globalrole/validator.go#L153) and adopting a similar approach.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
A couple of items:
- You need escalation checks for [GlobalRoleBindings](https://github.com/rancher/webhook/tree/109fe534b73bc47374dfd5f8a962d3f3d6545eb6/pkg/resources/management.cattle.io/v3/globalrolebinding) as well as GlobalRoles
- Please update the docs for the GlobalRole checks.
- The escalation logic is a bit flawed. The approach that we use needs to consider these permissions on their own since it gives permissions on all current workspaces and all future workspaces (besides local).

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator_test.go
Comment (raulcabello) [reply pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator_test.go:64]: added

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go
Comment (raulcabello) [reply pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go:64]: added

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/globalrole/validator.go]: I added the `fleetWorkspaceRules` to the `ClusterRulesFromRole` in the `GlobalRoleResolver`. This approach reuses the existing escalation check logic for the `inheritedClusterRoles`. By reusing the existing logic, the check is now included for both `GlobalRole` and `GlobalRoleBinding`. 
Please have a look and let me know what you think.

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator.go
Comment (MbolotSuse) [inline pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator.go:74]: I appreciate that you are trying to make this more testable, but it's very possible to write tests for these objects without resorting to a change like this. [Example Code](https://github.com/rancher/webhook/blob/54bba7ed560004f40cd277aef921bdcd07802ea6/pkg/resources/management.cattle.io/v3/globalrole/validator.go#L84), [Example Test](https://github.com/rancher/webhook/blob/54bba7ed560004f40cd277aef921bdcd07802ea6/pkg/resources/management.cattle.io/v3/globalrole/validator_test.go#L760). Since we don't need this to support testing this function, I would prefer that you don't add an interface here.

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator.go
Comment (MbolotSuse) [inline pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator.go:75]: Similar comment here - the interface is unnecessary, we can call out to the function directly.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Some minor issues and one overall larger issue regarding escalation checking.

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go:78]: I created 2 new resolvers: one for `FleetWorkspacePermissions.ResourceRules` and another for `FleetWorkspacePermissions.WorkspaceVerbs`

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator.go
Comment (raulcabello) [reply pkg/resources/rbac.authorization.k8s.io/v1/clusterrole/validator.go]: changed to follow the same approach as the GlobalRole test

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator.go
Comment (raulcabello) [reply pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator.go]: changed to follow the same approach as the GlobalRole validator test

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/globalrole/validator.go:138]: It has the `omitempty` json tag, which will set the slice to nil if empty. However, this tag works just if the resource is encoded to json. There are some cases where this encoding might not happen (e.g. kubectl). 
I added this check, and a unit test. Please notice I had to add the raw json in the unit test because the slice is converted to nil when encoding to json

File: pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go
Comment (MbolotSuse) [inline pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go:41]: Similar comment here about making the parameter order at minimum commented, and probably also looking into some better u/x here.

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go
Comment (MbolotSuse) [inline pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go:135]: I might be missing it, but I'm not sure how this test case differs from the last one. Would you mind taking a look at that?

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
A few concerns on the architecture/code structure, but the functionality/escalation checking looks much better.

File: pkg/resolvers/grbClusterResolver.go
Comment (raulcabello) [reply pkg/resolvers/grbClusterResolver.go]: what if we aggregate the three resolvers? I aggregated them in a new struct `GRBRuleResolvers`. That way we can reference them in the validator, and it's clear which rules are being resolved. Let me know what you think

File: pkg/resolvers/grbClusterResolver.go
Comment (raulcabello) [reply pkg/resolvers/grbClusterResolver.go]: that was a mistake, fixed

File: pkg/resources/management.cattle.io/v3/globalrole/validator.go
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/globalrole/validator.go]: see https://github.com/rancher/webhook/pull/348#discussion_r1576015559

File: pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/globalrolebinding/validator.go]: see https://github.com/rancher/webhook/pull/348#discussion_r1576015559

File: pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go
Comment (raulcabello) [reply pkg/resources/rbac.authorization.k8s.io/v1/clusterrolebinding/validator_test.go]: this was a left over. Error is tested in `TestAdmin_errors`. I removed it

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go:72]: I don't fully understand what you mean here. These rules will be checked with the `FleetWorkspaceRulesResolver`, which means it'll check if the user have enough permission in all workspace except local (by looking at the `InheritedFleetWorkspacePermissions.ResourceRules` of the `GlobalRoles` he currently has)

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go:80]: I don't understand what you mean here. Could you elaborate on this, please?

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go:90]: Comment added. It should be fine if we just use it to evaluate `InheritedFleetWorkspacePermissions.WorkspaceVerbs` rules with a `FleetWorkspaceRuleResolver`.
Alternatively, I could generate the resourceNames that'll contain all workspaces except local. That would mean listing all workspaces each time this is evaluated.

File: pkg/auth/globalrole.go
Comment (MbolotSuse) [reply pkg/auth/globalrole.go:72]: [Here](https://github.com/rancher/webhook/blob/4e47fcb77bd14e000b35fa486e9a1fd71f60994b/pkg/auth/globalrole.go#L51) is the relevant logic. Basically, even though RestrictedAdmin has no permissions in these fields, we need to give the appropriate permissions so that RA can give this role to others. But the RA doesn't have/doesn't use this field, so you need special logic here to handle that.

File: pkg/auth/globalrole.go
Comment (MbolotSuse) [reply pkg/auth/globalrole.go:84]: Sure - [Here](https://github.com/rancher/webhook/blob/4e47fcb77bd14e000b35fa486e9a1fd71f60994b/pkg/auth/globalrole.go#L51) is the relevant logic. Basically, even though RestrictedAdmin has no permissions in these fields, we need to give the appropriate permissions so that RA can give this role to others. But the RA doesn't have/doesn't use this field, so you need special logic here to handle that.

File: pkg/auth/globalrole.go
Comment (MbolotSuse) [reply pkg/auth/globalrole.go:90]: Thanks. Agree that this should be safe for our limited usage, and the comment should dissuade others from making bad changes here.

File: pkg/resolvers/grbClusterResolver.go
Comment (MbolotSuse) [reply pkg/resolvers/grbClusterResolver.go]: Yep, I think that this works. Since the struct is exported, I think it would be good to add comments above each field to specify what exactly that specific resolver is for, but overall I think that this structure works. Resolving.

**Review by JonCrowther — DISMISSED** [review]

**Review by JonCrowther — APPROVED** [review]

**Comment by samjustus** [comment]
[#42170](https://github.com/rancher/rancher/issues/42170)

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go]: done

File: pkg/resolvers/grbClusterResolver.go
Comment (raulcabello) [reply pkg/resolvers/grbClusterResolver.go:1]: renamed to GRBRuleResolver

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go:72]: done. I restricted the resources to just fleet resources, so RA cant grant core types in the local cluster. please have a look https://github.com/rancher/webhook/pull/348/commits/8525e3866f03bdb10c9d22556a98fa084678a7fe

File: pkg/auth/globalrole.go
Comment (raulcabello) [reply pkg/auth/globalrole.go:96]: done. see https://github.com/rancher/webhook/pull/348#discussion_r1639880143

File: pkg/auth/globalrole.go
Comment (JonCrowther) [inline pkg/auth/globalrole.go:72]: This exported function should have a comment describing what it's for.

Comment (raulcabello) [reply pkg/auth/globalrole.go:76]: you're right! done in https://github.com/rancher/webhook/pull/348/commits/57b1a9fe5620b0e9b32d2368caaad5834c9ec84d

**Review by JonCrowther — DISMISSED** [review]

**Review by JonCrowther — DISMISSED** [review]

**Review by MbolotSuse — COMMENTED** [review]
One NIT, but outside of that LGTM (After fixing merge conflict).

File: go.mod
Comment (raulcabello) [reply go.mod]: those were changed because I bumped the fleet version in order to have the ExternalRules field. This is not needed anymore since the k8s 1.29 PR was merged, which was also bumping the rancher version. No go.mod changes after fixing the conflicts

**Review by JonCrowther — APPROVED** [review]

**Review by MbolotSuse — APPROVED** [review]

**Pull request merged**
